drivers/overlay/overlay.go

```diff
@@ -1021,8 +1021,39 @@ func (d *Driver) Put(id string) error {
 	if _, err := ioutil.ReadFile(path.Join(dir, lowerFile)); err != nil && !os.IsNotExist(err) {
 		return err
 	}
-	if err := unix.Unmount(mountpoint, unix.MNT_DETACH); err != nil && !os.IsNotExist(err) {
-		logrus.Debugf("Failed to unmount %s overlay: %s - %v", id, mountpoint, err)
+
+	unmounted := false
+
+	if d.options.mountProgram != "" {
+		// Attempt to unmount the FUSE mount using either fusermount or fusermount3.
+		// If they fail, fallback to unix.Unmount
+		for _, v := range []string{"fusermount3", "fusermount"} {
+			err := exec.Command(v, "-u", mountpoint).Run()
+			if err != nil && !os.IsNotExist(err) {
+				logrus.Debugf("Error unmounting %s with %s - %v", mountpoint, v, err)
+			}
+			if err == nil {
```

Comment:
Seems like you are doing a lot of work if os.IsNotExists(err) is true, Should we set unmounted=trued under those circumstances as well, it is just going to cause everything below to fail multiple times?

Reply:
the `os.IsNotExists(err)` should not be treated differently than the case `err != nil` as the mount is still not freed.

When we could not unmount using `fusermount|fusermount3` then we fallback to syncfs+unmount(2) that frees the mount and the `syncfs` makes sure any pending request is flushed to the file system

Reply:
Is the err IsNotExists referring to the Mount point does not exists, won't it still not exists?  So the rest of the calls will fail as well.